api-docs: Edit message endpoint and related API changelog entry revisions.

laurynmm opened the pull request:
Part of 7.0 API changelog audit; see [this CZO conversation](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/7.2E0.20API.20changelog.20audit/near/1573165) for more context.

Updates API changelog entry and related parts of API docs for: **Feature levels 172, 162 and 159**. 

**Notes**:
- All of these feature levels were documented on the `PATCH /messages/{message_id}` endpoint, so I felt like that page's updates would work best in one PR.
- Each commit focuses on a specific feature level's API documentation changes.
- Reorders and updates related **Changes** notes so that they go from most recent to least recent.
- Adds some more details and links to the main description and special error response description in `PATCH /messages/{message_id}`
- Consolidates some API feature level entries and revises text there a bit.

---

**Screenshots and screen captures:**

<details>
<summary>Edit a message: main endpoint description</summary>

[Current documentation](https://chat.zulip.org/api/update-message)
![Screenshot from 2023-05-22 21-23-21](https://github.com/zulip/zulip/assets/63245456/b78122a7-79b4-4cc0-ad9f-213d18bc606c)
</details>
<details>
<summary>Edit a message: special error response</summary>

[Current documentation](https://zulip.com/api/update-message#response)
![Screenshot from 2023-05-22 21-00-27](https://github.com/zulip/zulip/assets/63245456/e90d16cd-0b8a-42dd-92f6-c1dacc1cd489)
</details>
<details>
<summary>API changelog entries</summary>

[Current documentation](https://chat.zulip.org/api/changelog#changes-in-zulip-70)
#### Feature level 172
![Screenshot from 2023-05-22 21-03-23](https://github.com/zulip/zulip/assets/63245456/201189ac-db17-413a-99f4-38981360aa9d)
#### Feature level 162
![Screenshot from 2023-05-22 21-05-21](https://github.com/zulip/zulip/assets/63245456/ec6278ea-60a5-4d65-af9b-46c9b98f83bd)
#### Feature level 159
![Screenshot from 2023-05-22 21-05-34](https://github.com/zulip/zulip/assets/63245456/49126f1c-95f3-434d-8021-5d32324e4883)
</details>
<details>
<summary>Register response: search for feature level 162</summary>

[Current documentation](https://chat.zulip.org/api/register-queue)
![Screenshot from 2023-05-22 21-07-36](https://github.com/zulip/zulip/assets/63245456/710038ba-ab1d-4d69-a670-5301adf47587)
</details>
<details>
<summary>Get events: search for feature level 162</summary>

[Current documentation](https://chat.zulip.org/api/get-events#realm-update_dict)
![Screenshot from 2023-05-22 21-09-56](https://github.com/zulip/zulip/assets/63245456/11e74d0d-e21f-49bb-859f-aeb5cc9bb70d)
</details>
<details>
<summary>Register response: search for feature level 159</summary>

[Current documentation](https://chat.zulip.org/api/register-queue)
![Screenshot from 2023-05-22 21-12-30](https://github.com/zulip/zulip/assets/63245456/95754726-f167-48f0-9238-44ed3dcea630)
![Screenshot from 2023-05-22 21-12-41](https://github.com/zulip/zulip/assets/63245456/4099ee35-8536-4159-bb70-4e5c7f04a5fc)
</details>
<details>
<summary>Get events: search for feature level 159</summary>

[Current documentation](https://chat.zulip.org/api/get-events#realm-update_dict)
![Screenshot from 2023-05-22 21-14-36](https://github.com/zulip/zulip/assets/63245456/72c74269-aae1-46ba-b8d4-4a7446a4bbff)
![Screenshot from 2023-05-22 21-14-47](https://github.com/zulip/zulip/assets/63245456/ef5cf5cb-e650-4e56-a2d2-17bba6904515)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -6892,35 +6896,48 @@ paths:
+        had permission to move messages between streams did not have any time limit
+        restrictions when doing so.
+
+        As of Zulip 7.0 (feature level 159), the realm settings
+        `move_messages_between_streams_policy` and `edit_topic_policy` always
+        determine if a user has permission to move messages between streams or edit
+        topics. Previously, editing streams and topics of messages was forbidden
+        if `allow_message_editing` was `false`, regardless of the above policies.
```
Here also, I think putting it in "Before" form (like in the existing text) rather than "As of" form ends up being clearer.

In addition to the general points in the other comment, the "always" here makes me a bit nervous because it says there's no *other* constraints that apply.

laurynmm replied:
I had removed that "always" and must have not pushed that most recent version. The current revisions have gone back to the "Before" form for this note, though I did expand the text to note these are realm settings.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -4066,29 +4067,31 @@ paths:
+                                        not affect moderators and administrators.
 
-                                        Will not be 0. A `null` value means no limit: messages can be moved
+                                        Will not be 0. A `null` value means no limit, and messages can be moved
```
nit: perhaps
```suggestion
                                        Will not be 0. A `null` value means no limit, so that messages can be moved
```

gnprice commented on api_docs/changelog.md:
```diff
@@ -236,16 +239,15 @@ format used by the Zulip server that they are interacting with.
+  [`GET /events`](/api/get-events):
+  Nobody added as an option for the realm settings `edit_topic_policy`
+  and `move_messages_between_streams_policy`.
+* [`PATCH /messages/{message_id}`](/api/update-message): Permission
+  to edit the stream and/or topic of messages no longer depends on the
+  realm setting `allow_message_editing`.
```
(By contrast, this way around makes perfect sense here in the API changelog. Focusing on the new state rather than the old is the normal thing to do in a changelog; so I guess you could say that I think of a "Changes" section in the rest of the API docs as more like a "reverse changelog", where the reader starts from understanding the new state and then wants to precisely understand the old states.)

laurynmm replied:
That's helpful context for me when reviewing and revising these. Thanks!

timabbott replied:
Yeah this might be the type of idea that we should take the opportunity to write in our API docs documentation, which I think generally needs a pass for the fact that the majority of work on our API docs is documenting new changes, not new endpoints, these days.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -7057,30 +7074,35 @@ paths:
                             "total_messages_in_topic": 5,
                           }
                         description: |
-                          A special failed JSON response for when the user has permission to move
+                          A special failed JSON response (`"code": "MOVE_MESSAGES_TIME_LIMIT_EXCEEDED"`)
```
Good idea. The way we render these example responses gets pretty confusing once their descriptions are more than a line or two long:
![image](https://github.com/zulip/zulip/assets/28173/3d6e9f7c-d464-4eaa-bc68-184f4f32ba36)

in that it looks very much like the paragraph "A special failed JSON response …" describes the example immediately *above* it, rather than the example far below. (When they're all just a line long, the colon at the end of the line does an OK job at disambiguating.)

Ideally we'd see about fixing that at some point. But in the meantime, saying `"code": "MOVE_MESSAGES_TIME_LIMIT_EXCEEDED"`` here should do a lot to help people determine quickly that it's not about the example above, so they can look again and puzzle out that it's about the example far below.

gnprice reviewed (commented):
Thanks for cleaning these up! Small comments below.

laurynmm commented:
**Updates**:
- Switched the feature level 159 **Changes** notes in the `/api/update-message` endpoint main description back to "Before" format, as suggested by review feedback above.
- Added a commit for expanding that same main endpoint description to have a list of the relevant realm settings for reference.

**Updated screenshots**:

<details>
<summary>Edit a message: main description</summary>

![Screenshot from 2023-05-23 17-47-21](https://github.com/zulip/zulip/assets/63245456/e1585153-0d1e-475c-9199-ec5f9106c377)
</details>
<details>
<summary>Get events: feature level 162 notes</summary>

![Screenshot from 2023-05-23 17-49-13](https://github.com/zulip/zulip/assets/63245456/e716930b-346a-40e4-aa74-89ea90e57ef1)
</details>
<details>
<summary>Register queue: feature level 162 notes</summary>

![Screenshot from 2023-05-23 17-50-15](https://github.com/zulip/zulip/assets/63245456/a15677af-f63d-47fc-b6ab-27b4efc6c5d9)
</details>

timabbott commented:
This looks great, merged, thanks @laurynmm!

gnprice commented:
All looks great, thanks!
